osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerUserModSelectOverlay.cs

```diff
@@ -0,0 +1,229 @@
+        /// </summary>
+        private void onUserModsChanged(MultiplayerRoomUser user)
+        {
+            if (user.Equals(client.LocalUser))
+                updateFromOnlineState();
+        }
+
+        /// <summary>
+        /// Responds to changes in the currently-selected playlist item to process changes to <see cref="MultiplayerPlaylistItem.BeatmapID"/>,
+        /// <see cref="MultiplayerPlaylistItem.RulesetID"/>, <see cref="MultiplayerPlaylistItem.AllowedMods"/>, or <see cref="MultiplayerPlaylistItem.Freestyle"/>.
+        /// </summary>
+        private void onPlaylistItemChanged(MultiplayerPlaylistItem item)
+        {
+            if (item.ID == client.Room?.Settings.PlaylistItemId)
+                updateFromOnlineState();
+        }
+
+        /// <summary>
+        /// Responds to changes in the active playlist item that could occur as a result of it being deleted or the room's queue mode having been changed.
+        /// </summary>
+        private void onSettingsChanged(MultiplayerRoomSettings settings)
+        {
+            if (settings.PlaylistItemId != lastPlaylistItemId)
+            {
+                updateFromOnlineState();
+                lastPlaylistItemId = settings.PlaylistItemId;
+            }
+        }
+
+        /// <summary>
+        /// Tracks any changes to the selected mods.
+        /// </summary>
+        private void onSelectedModsChanged(ValueChangedEvent<IReadOnlyList<Mod>> mods)
+        {
+            hasPendingChanges = true;
+            modSettingChangeTracker?.Dispose();
+            modSettingChangeTracker = new ModSettingChangeTracker(mods.NewValue);
+            modSettingChangeTracker.SettingChanged += _ => hasPendingChanges = true;
+        }
+
+        /// <summary>
+        /// Processes changes to the selected mods to notify the server of any changes at a debounced rate.
+        /// </summary>
+        private void processPendingChanges()
+        {
+            if (!hasPendingChanges)
+                return;
+
+            if (Time.Current - lastChangeTime < update_debounce_time)
+                return;
+
+            IDisposable clientAuthority = createClientAuthorityToken();
+
+            client.ChangeUserMods(SelectedMods.Value)
+                  .FireAndForget()
+                  .ContinueWith(_ => clientAuthority.Dispose());
+
+            hasPendingChanges = false;
+            lastChangeTime = Time.Current;
+        }
+
+        /// <summary>
+        /// Creates a token that causes this overlay to take complete authority over all changes to the selected mods,
+        /// preventing server-side changes from being applied (<see cref="isClientAuthoritative"/>).
+        /// </summary>
+        /// <returns>
+        /// The token, which should be disposed in order to restore the server's authority.
+        /// </returns>
+        private IDisposable createClientAuthorityToken()
```

Comment:
i'm gonna need an in-depth explanation of what's going on here because my initial reactions are "what on earth is this" followed by "why though i don't recall seeing something like this ever before in multiplayer code" and a complete brain shutdown and the one sentence about it in the OP isn't helping